src/stavenote.ts

```diff
@@ -164,11 +164,64 @@ export class StaveNote extends StemmableNote {
       });
     }
 
-    const voices = notesList.length;
+    let voices = notesList.length;
+    if (voices < 2) return true;
 
     let noteU = notesList[0];
-    const noteM = voices > 2 ? notesList[1] : undefined;
-    let noteL = voices > 2 ? notesList[2] : notesList[1];
+    let noteM = undefined;
+    let noteL = undefined;
+    let draw = noteU.note.render_options.draw ? noteU.note.render_options.draw : true;
+
+    if (voices === 2) {
+      // 2 voices & notesList[0] hidden, return
+      if (!draw) return true;
+      noteL = notesList[1];
+      draw = noteL.note.render_options.draw ? noteL.note.render_options.draw : true;
+      // 2 voices & notesList[1] hidden, return
+      if (!draw) return true;
```

Comment:
There are a lot of these `if (!draw) return true`, and it's pretty hard to reason about. Can you group or restructure them so there are fewer code paths to navigate?

Reply:
This is the function shifting notes if they overlap. The calculation was ignoring that some of the notes might be hidden and hidden botes do not overlap.
I have put a comment on each return. The one above tells tha we have two voices but one of them is hidden, as a consquence no shift is required and we can return.

Reply:
I changed the approach and now there is only one return and I believe it is much clearer. Thanks for the comment I got lost trying to follow the original approach.